### ISSUE TITLE
Examples or reference materials.

### ISSUE THREAD

**Issue by hally9k** [cartant/rxjs-marbles#25]
Firstly, many thanks for some awesome work here.

I have just spent the morning getting everything working with our Jest setup. Struggled a bit TBH. I spent a lot of time hopping between blogs and docs trying to piece together a decent real world test example.

My output is here: https://github.com/hally9k/rxjs-marbles-jest-example

This issue is really 2 questions:

1. Should we add some examples like this to the source or links to the readme to help future noobs like myself?

2. Does my output referenced serve as good example or have I missed a better way to do this? (It doesn't feel super clean the way that I am mocking delay...) 🤔

**Comment by cartant** [comment]
Regarding the first question: yes, in [light of this issue](https://github.com/cartant/rxjs-marbles/issues/24), from yesterday, I've added a task to my TODO list to improve the documentation. It's grown over time and could do with being reorganised. And, yes, the examples could be improved.

As for the second, I will check out the example you've referenced a little later.

Thanks for raising the issue.

**Comment by hally9k** [comment]
@cartant Did you ever get round to looking at this for me?

**Comment by cartant** [comment]
I've had a look at your example. Thanks for the effort that you've put into it.

However, I think tests that involve schedulers should use the `TestScheduler` exposed by the context - via `m.scheduler` - rather than mocking the operators themselves. Your example uses an implicit scheduler, as `delay` [defaults to using the `AsyncScheduler`](https://github.com/ReactiveX/rxjs/blob/5.5.8/src/operators/delay.ts#L53).

When testing with schedulers you have two options:

* pass the scheduler explicitly; or
* use the context's `bind` method to [bind all schedulers to the test's `TestScheduler` instance](https://github.com/cartant/rxjs-marbles#dealing-with-deeply-nested-schedulers).

I would do something like this:

```js
describe('Delayer', () => {
    it(
        'should complete when it receives a payload with a delay value of 4 seconds or greater.',
        marbles(m => {

            const values = {
                a: { value: 'alpha', delay: 10 },
                b: { value: 'beta', delay: 20 },
                c: { value: 'charlie', delay: 30 },
                d: { value: 'delta', delay: 40 }
            }

            const producer = m.hot('^a-b--c---d|', values)
            const expected = m.hot('^-a--b---c|', values)

            m.bind();
            const observable = delayer(producer)
            m.expect(observable).toBeObservable(expected)
        })
    )
})
```

The clear benefit is that the expected marble diagram includes the delays. (Note that you will need to reduce your `delay` values, as each character in the diagram counts as 10 frames.)

With `delayer` composed like this (note that to simplify the expected values, `x` is not mapped to `x.value`):

```js
module.exports = subject =>
    Observable.from(subject)
        .takeWhile(x => x.delay < 40)
        .concatMap(x => Observable.of(x).delay(x.delay))
        .do(console.log)
```

I've created a `v6` branch that includes the changes necessary to support RxJS version 6. At some stage around the release date for version 6, I will rearrange the `README.md` and will create some working examples within an `examples` directory in the repo. I'll add some basic examples and if others want to submit PRs with further examples, they would be welcome to do so.

**Comment by hally9k** [comment]
@cartant Thanks so much for taking the time, this was SUPER helpful.
One question: If I want to test this delayer with the original larger delay values can I change the amount of frames that each character in the diagram represents to 100 or 1000? A frame resolution setting I suppose 🤔

**Comment by cartant** [comment]
I don't think you will be able to change it.

The maximum frame of 750 is [a constant](https://github.com/ReactiveX/rxjs/blob/5.5.8/src/testing/TestScheduler.ts#L10) and it's passed to the `VirtualTimeScheduler` in the `TestScheduler` [constructor](https://github.com/ReactiveX/rxjs/blob/5.5.8/src/testing/TestScheduler.ts#L27). And the `frameTimeFactor` is a protected (in TypeScript) [static property](https://github.com/ReactiveX/rxjs/blob/5.5.8/src/scheduler/VirtualTimeScheduler.ts#L7) in `VirtualTimeScheduler`. So changing frames per character is not really something that's supported/anticipated.

Around the time I published `rxjs-marbles`, OJ Kwon - he's on the RxJS core team - published [`rx-sandbox`](https://github.com/kwonoj/rx-sandbox). It has it's own implementation of `TestScheduler` which supports specifying - IIRC - the number of frames per character. You might want to check it out, if that's a must-have feature for you.

**Comment by hally9k** [comment]
Oh, so we can't test anything that is more than 3/4 of a second delayed? What is the motive for having this threshold?

**Comment by hally9k** [comment]
Thanks for the references, I have a bit reading to do no doubt. I'm a little out of touch with the movement in the RxJS space, clearly

In our project we have some timings that are part of the characteristics under test. If the 75 char limit is an arbitrary limit based on assumed use cases maybe we could make it configurable, or is there some technical reason we would want to limit it to a specific length?

**Comment by cartant** [comment]
There's no reason why it shouldn't be more configurable. It's just that it was built to test the RxJS library and it does that. The authors didn't need it to do anything else, so it doesn't - YAGNI and all that.

I'll make an note to have a closer look at the internals, later, to see if there's any way I can just clobber it. I doubt it would be a bigger hack than `m.bind()` :smile:

**Comment by hally9k** [comment]
Nice! Let me know how you get on ;) Thanks again!

**Comment by cartant** [comment]
Actually, it'd be pretty simple to add this to `rxjs-marbles`. I'll likely do it tonight - i.e. in 12 hours or so.

**Comment by hally9k** [comment]
@cartant So I've been jamming a bit with `rx-sandbox` like you said and the test scheduler has the features I want but there's no good way to inject the test scheduler into the operaters when their in the client code. I read your bind function on the context to get a feel for how you're doing it (didn;t fully grok it TBH) and noticed that you can pass in an array of schedulers to the bind function. Have you tried passing the `rx-sandbox` test scheduler into the `rxjs-marbles` bind function?

**Comment by cartant** [comment]
Passing OJ's `TestScheduler` to `bind` won't help. `bind` monkey patches the `now` and `schedule` methods of the default  schedulers to call the `TestScheduler` held by the `Context`.

However, you'd want to bind the default schedulers to the instance of OJ's scheduler.

I guess you could try passing and instance of OJ's scheduler to the `Context`, but to do that you'd be really digging down into the implementation of `rxjs-marbles` - which really tries to make the `Context` the primary/ only API surface.

Regarding `bind`, I [saw this comment](https://reactivex.slack.com/archives/C0PQCKDV1/p1511297528000474) from Jay Phelps on Slack:

> I think I'm going to hell. :fire:
> 
> ```
> import { async } from 'rxjs/scheduler/async';
> async.schedule = testScheduler.schedule.bind(testScheduler);
> ```

And decided that - whilst it my see me, too, also going to hell - it would make my life a lot easier when it comes to deeply nested schedulers. `bind` just monkey patches `schedule` and  `now` and returns an array that contains the original functions - so that they can be restored when the context is torn down.

**Comment by cartant** [comment]
What I would really like is to be able to specify the scheduler in the call to `subscribe` - as suggested by Paul Taylor in [this comment](https://github.com/ReactiveX/rxjs/issues/2935#issuecomment-336681001),

Being able to do that might save me from eternal damnation. :fire: :smile: :fire:

**Comment by hally9k** [comment]
@cartant I may need an invitation to see that slack comment? I'm not in that slack community.

**Comment by cartant** [comment]
If you want an invite, just ask Tracy Lee. See [this tweet](https://twitter.com/ladyleet/status/938855641556406273) or you could email her: tracy@thisdot.​co

**Comment by hally9k** [comment]
Oh you did paste the whole thing...? Not even slightly clearer though I'm afraid, I clearly require quite a bit of study of the RxJS innards. That snippet looks pretty crazy hacky though 😂
Anyway, I've reverted to passing mock delay values in for my tests for now which is a compromise.
If you manage to clobber this char limit I'll be able to test it properly tomorrow which will be waaay better for this particular application.
The frame resolution thing though, that would be awesome. Thanks for all your help today, I really appreciate it.

**Comment by cartant** [comment]
I've published 2.4.1 which adds a `reframe` method to the context, It allows the specification of the frames per character and the maximum number of frames.

There's [a fixture](https://github.com/cartant/rxjs-marbles/blob/v2.4.1/fixtures/mocha/passing-spec.ts#L229-L239) that demonstrates the usage:

```ts
it("should support reframing", marbles((m) => {

    m.reframe(100, 10000);

    const duration = m.time("--|");
    expect(duration).to.equal(200);

    const source =   m.cold("--(a|)");
    const expected = m.cold("----(a|)");
    m.expect(source.delay(duration, m.scheduler)).toBeObservable(expected);
}));
```

The clobbering is restored after the context is torn down, so the reframing is local the each test.

If you always want to use a particular framing configuration, you could create a module and could export a wrapped `marbles` function in a manner similar to that used for the framework-specific `marbles` functions in `rxjs-marbles`:

```ts
import { marbles as _marbles } from "rxjs-marbles/jest";

export function marbles(func: (m: Context, ...rest: any[]) => any): any {
    return _marbles((m: Context, ...rest: any[]) => {
        m.reframe(100, 10000);
        return func(m, ...rest);
    });
}
```

Closing this. When RxJS v6 is released, I'll release a version of `rxjs-marbles` that supports it and that version will include examples. At that stage, contributions of further example will be welcome.

**Comment by hally9k** [comment]
You sir, are an absolute legend! 🎉
